rpmio/rpmkeyring.c

```diff
@@ -153,37 +155,72 @@ rpmPubkey rpmPubkeyNew(const uint8_t *pkt, size_t pktlen)
     return key;
 }
 
-static rpmPubkey rpmPubkeyNewSubkey(pgpDigParams pgpkey)
+static rpmPubkey rpmPubkeyNewSubkey(rpmPubkey primarykey, pgpDigParams pgpkey)
 {
     rpmPubkey key = new rpmPubkey_s {};
-    /* Packets with all subkeys already stored in main key */
+    /* Packets with all subkeys already stored in primary key */
+    key->primarykey = primarykey;
     key->pgpkey = pgpkey;
     key->keyid = key2str(pgpDigParamsSignID(pgpkey));
     key->nrefs = 1;
     return key;
 }
 
-rpmPubkey *rpmGetSubkeys(rpmPubkey mainkey, int *count)
+rpmPubkey *rpmGetSubkeys(rpmPubkey primarykey, int *count)
 {
     rpmPubkey *subkeys = NULL;
     pgpDigParams *pgpsubkeys = NULL;
     int pgpsubkeysCount = 0;
     int i;
 
-    if (mainkey && !pgpPrtParamsSubkeys(mainkey->pkt.data(), mainkey->pkt.size(),
-			mainkey->pgpkey, &pgpsubkeys, &pgpsubkeysCount)) {
+    wrlock lock(primarykey->mutex);
+
+    if (primarykey && !pgpPrtParamsSubkeys(primarykey->pkt.data(), primarykey->pkt.size(),
+			primarykey->pgpkey, &pgpsubkeys, &pgpsubkeysCount)) {
 
 	/* Returned to C, can't use new */
 	subkeys = (rpmPubkey *)xmalloc(pgpsubkeysCount * sizeof(*subkeys));
-	for (i = 0; i < pgpsubkeysCount; i++)
-	    subkeys[i] = rpmPubkeyNewSubkey(pgpsubkeys[i]);
+	for (i = 0; i < pgpsubkeysCount; i++) {
+	    subkeys[i] = rpmPubkeyNewSubkey(primarykey, pgpsubkeys[i]);
+	    primarykey->nrefs++; /* can't use rpmPubkeyLink due to wrlock */
+	}
 	free(pgpsubkeys);
     }
     *count = pgpsubkeysCount;
 
     return subkeys;
 }
 
+rpmPubkey rpmPubkeyPrimarykey(rpmPubkey key)
+{
+    rpmPubkey primarykey = NULL;
+    if (key) {
+	wrlock lock(key->mutex);
+	if (key->primarykey == NULL) {
+	    primarykey = key;
+	    primarykey->nrefs++; /* can't use rpmPubkeyLink due to wrlock */
+	} else {
+	    primarykey = key->primarykey;
+	    rpmPubkeyLink(primarykey);
```

Comment:
Make that `primarykey = rpmPubkeyLink(key->primarykey)` to make it clear what happens there. You should think of fooLink() as returning a new reference rather than something just does ++ internally.

Reply:
Done.